File: python/88.合并两个有序数组.py
#
# @lc app=leetcode.cn id=88 lang=python3
#
# [88] 合并两个有序数组
#

# @lc code=start
class Solution:
    # 逆向双指针，不用多复制一个数组
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        p1 = m - 1
        p2 = n - 1
        tail = m + n - 1
        while p1 >= 0 or p2 >= 0:
            if p1 <= -1:
                nums1[tail] = nums2[p2]
                p2 -= 1
            elif p2 <= -1:
                nums1[tail] = nums1[p1]
                p1 -= 1
            elif nums1[p1] < nums2[p2]:
                nums1[tail] = nums2[p2]
                p2 -= 1
            else:
                nums1[tail] = nums1[p1]
                p1 -= 1
            tail -= 1


    # 双指针
    def merge2(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        sorted = [0] * (m + n)
        cur = 0
        p1 = 0
        p2 = 0
        while p1 < m or p2 < n:
            if p1 == m:
                sorted[cur] = nums2[p2]
                p2 += 1
            elif p2 == n:
                sorted[cur] = nums1[p1]
                p1 += 1
            elif nums1[p1] < nums2[p2]:
                sorted[cur] = nums1[p1]
                p1 += 1
            else:
                sorted[cur] = nums2[p2]
                p2 += 1
            cur += 1
        nums1[:] = sorted

    # 调用api，不满足m + n时间要求
    def merge1(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        nums1[m:] = nums2
        nums1.sort()
# @lc code=end

File: cpp/88.合并两个有序数组.cpp
/*
 * @lc app=leetcode.cn id=88 lang=cpp
 *
 * [88] 合并两个有序数组
 */
#include <algorithm>
#include <iostream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
// @lc code=start
class Solution
{
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n)
    {
        method2(nums1, m, nums2, n);
    }

    // 逆向双指针
    void method2(vector<int>& nums1, int m, vector<int>& nums2, int n)
    {
        int tail = m + n - 1;
        int p1   = m - 1;
        int p2   = n - 1;
        while (p1 >= 0 || p2 >= 0) {
            int cur = 0;
            if (p1 < 0) {
                cur = nums2[p2--];
            }
            else if (p2 < 0) {
                cur = nums1[p1--];
            }
            else if (nums1[p1] > nums2[p2]) {
                cur = nums1[p1--];
            }
            else {
                cur = nums2[p2--];
            }
            nums1[tail--] = cur;
        }
    }

    // api法
    void method1(vector<int>& nums1, int m, vector<int>& nums2, int n)
    {
        for (int i = 0; i < n; i++) {
            nums1[m + i] = nums2[i];
        }

        sort(nums1.begin(), nums1.end());
    }
};
// @lc code=end

int main()
{
    vector<int> v1 = {1, 2, 3, 0, 0, 0};
    vector<int> v2 = {2, 5, 6};
    (Solution()).merge(v1, 6, v2, 3);
}
